File: tabulation/kotlin/canSum.kt
fun canSum(targetSum: Int, numbers: Array<Int>): Boolean {
    val table: Array<Boolean> = Array(targetSum + 1) { false }
    table[0] = true

    for (i: Int in (0..targetSum)) {
        if (table[i]) {
            numbers.forEach {
                if (i + it <= targetSum) table[i + it] = true
            }
        }
    }

    return table[targetSum]
}

fun main(): Unit {
    println(canSum(303, arrayOf(5, 6, 7)))
}

File: tabulation/javascript/canSum.js
// m = targetSum
// n = numbers.length
// O(nm) time
// O(m) space
function canSum(targetSum, numbers) {
    const table = Array(targetSum + 1).fill(false)
    table[0] = true

    for (let i = 0; i <= targetSum; i++) {
        if (table[i])
            for (let num of numbers)
                table[i + num] = true
    }

    return table[targetSum]
}

console.log(canSum(514, [5, 3, 8, 12, 11]))